src/storage/src/render/upsert.rs

```diff
@@ -222,32 +295,104 @@ where
-            key_indices_sorted.sort_unstable();
             let key_indices_map = upsert_envelope
                 .key_indices
                 .iter()
                 .enumerate()
                 .map(|(idx, value_idx)| (*value_idx, idx))
                 .collect();
             let mut kdv = DatumVec::new();
+            // this is a map of (decoded key) -> (decoded_value). We store the
+            // latest value for a given key that way we know what to retract if
+            // a new value with the same key comes along
+            let mut current_values = None;
+
+            let mut initial_values_multiset: HashMap<(Row, Row), Diff> = HashMap::default();
+            move |data_input, previous_input, output| {
+                if previous_token.is_some() {
+                    assert!(current_values.is_none());
+                    // Hydrate the `current_values` map from the previous state of the collection.
+                    // We can't just insert things into the `current_values` map directly, since
+                    // we might in general have non-one multiplicities due to Persist being behind on compaction.
+                    // Thus, we use `initial_values_multiset` to keep track of how many of each record we've seen.
+                    //
+                    // At the end of reading the entire previous input, `initial_values_multiset` must have exactly one of each record,
+                    // and furthermore, each key must be unique. We validate this property for sanity's sake, and build `current_values`.
+                    //
+                    // TODO[btv] This algorithm has the potential to use unbounded space if we can't make any assumptions
+                    // about Persist's level of compaction or the order in which it returns updates.
+                    // See the detailed discussion [here](https://materializeinc.slack.com/archives/C01CFKM1QRF/p1659063332027139).
+                    // This is thought to be fine for our current purposes, but we will need to rethink it once the Persist team has
+                    // thought harder about what guarantees/APIs they want to offer relating to bounded-memory consolidation.
+                    //
+                    // TODO[btv] Note that we're only looking at `Ok` records here -- the `Err` entries go directly to the error
+                    // stream during source rendering before we ever get here. That means that if an `Err` made it into the
+                    // persisted state, and we restart before it's retracted, the user will never have a chance to retract it
+                    // and the source will be permanently dead. In my opinion, this should be considered a bug and fixed before GA,
+                    // but fixing it will require a bit of surgery, because right now the type of errors we store in persist don't
+                    // allow us to recover the original key and value separately.
```

Comment:
This seems maybe more important than an internal note. It sounds like this means that e.g. decoding errors are fatal for an upsert source (for now). It sounds like a legit limitation, but imo we should file something, and probably also mention clearly nearer the definition of some `fn upsert`.

Reply:
Filed an issue.